File: Depth-First Search/Graph/LET_no2246_Longest Path With Different Adjacent Characters.py
class Solution:
    def longestPath(self, parent: List[int], s: str) -> int:
        children = [[] for _ in range(len(s))]
        for i, p in enumerate(parent):
            if p >= 0:
                children[p].append(i)
            
        self.res = 0
        self.helper(children, 0, s)
        return self.res
    
    def helper(self, children, cur, s):
        temp = [0, 0]
        for child in children[cur]:
            length = self.helper(children, child, s)
            if s[child] != s[cur]:
                temp.append(length)
        
        temp.sort(reverse=True)
        self.res = max(self.res, 1 + sum(temp[:2]))
        return 1 + temp[0]
        

File: Depth-First Search/Graph/LET_no2246_Longest Path With Different Adjacent Characters.cpp
class Solution {
public:
    int longestPath(vector<int>& parent, string s) {
        int n = parent.size();
        vector<vector<int>> children(n);
        for (int i = 1; i < n; i++)
            children[parent[i]].push_back(i);

        int res = 1;
        dfs(0, children, s, res);
        return res;
    }

    int dfs(int cur_node, vector<vector<int>>& children, string& s, int& res) {
        int longest = 0, second_long = 0;
        for (int child_node : children[cur_node]) {
            int child_longest = dfs(child_node, children, s, res);
            
            if (s[cur_node] == s[child_node])
                continue;
            
            if (child_longest > longest) {
                second_long = longest;
                longest = child_longest;
            }
            else if (child_longest > second_long)
                second_long = child_longest;
        }

        res = max(res, longest + second_long + 1);            
        return longest + 1;
    }
};